src/shared_library.cc

```diff
@@ -242,4 +242,92 @@ SharedLibrary::GetEntrypoint(
   return Status::Success;
 }
 
+Status
+SharedLibrary::AddAdditionalDependencyDir(
+    const std::string& additional_path, std::wstring& original_path)
+{
+#ifdef _WIN32
+  const std::wstring PATH(L"Path");
+
+  DWORD len = GetEnvironmentVariableW(PATH.c_str(), NULL, 0);
+  if (len > 0) {
+    original_path.resize(len);
+    GetEnvironmentVariableW(PATH.c_str(), &original_path[0], len);
+  } else {
+    return Status(Status::Code::INTERNAL, "PATH variable is empty");
+  }
+
+  LOG_VERBOSE(1) << "Environment before extending PATH: "
+                 << std::string(original_path.begin(), original_path.end());
+
+  std::wstring updated_path_value =
+      std::wstring(additional_path.begin(), additional_path.end());
+  updated_path_value += original_path;
+
+  if (!SetEnvironmentVariableW(PATH.c_str(), updated_path_value.c_str())) {
+    LPSTR err_buffer = nullptr;
+    size_t size = FormatMessageA(
+        FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |
+            FORMAT_MESSAGE_IGNORE_INSERTS,
+        NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+        (LPSTR)&err_buffer, 0, NULL);
+    std::string errstr(err_buffer, size);
+    LocalFree(err_buffer);
+    return Status(
+        Status::Code::INTERNAL,
+        "failed to append user-provided directory to PATH " + errstr);
+  }
+
+  if (LOG_VERBOSE_IS_ON(1)) {
+    std::wstring path_after;
+    len = GetEnvironmentVariableW(PATH.c_str(), NULL, 0);
+    if (len > 0) {
+      path_after.resize(len);
+      GetEnvironmentVariableW(PATH.c_str(), &path_after[0], len);
+    }
+    LOG_VERBOSE(1) << "Environment after extending PATH: "
+                   << std::string(path_after.begin(), path_after.end());
+  }
+#else
+  LOG_WARNING
+      << "The parameter \"additional-dependency-dirs\" has been specified but "
+         "is not supported for Linux. It is currently a Windows-only feature. "
+         "No change to the environment will take effect.";
+#endif
+  return Status::Success;
+}
+
+Status
+SharedLibrary::RemoveAdditionalDependencyDir(std::wstring& original_path)
+{
+#ifdef _WIN32
+  const std::wstring PATH(L"Path");
+  if (!SetEnvironmentVariableW(PATH.c_str(), original_path.c_str())) {
```

Comment:
What happen if `original_path` value will be `NULL`?

Reply:
That wouldn't happen. In the calling function we initialize `original_path` with the default constructor--which will initialize it to an empty string. We check whether `original_path` is empty before calling this function and we do not set it to NULL anywhere in between.

Reply:
Have you tried it with an empty string?

Reply:
I'm saying it wouldn't make it in here if it were empty: https://github.com/triton-inference-server/core/pull/400/files#diff-c3cafbdf0ba28838ed1f0c9407e4cb16f33bc18f8274b326114fc3a6bbbbbf81R203

Reply:
I would like to see this scenario tested because of the definition in the method that been chosen. 
https://learn.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-setenvironmentvariablew
> If this parameter is NULL, the variable is deleted from the current process's environment.

Reply:
As an update, based on Guan's suggestion that we continue when the PATH variable is empty instead of returning an error, there is now a valid (albeit very rare) scenario where `original_path` can be empty. This is an scenario I was able to test locally and can confirm that passing an empty string will result in `Path=` ({empty string} as it would have been before), whereas, passing `NULL` is distinct and will completely clear `Path` from the list of variables. Therefore, my argument about never being able to pass `NULL`, the real concern here, still applies.